### ISSUE TITLE
optimizations for ARM

### ISSUE THREAD

**Issue by iit2014128** [githubharald/CTCWordBeamSearch#40]
Has anyone tried this algorithm on ARM Architecture? It is taking very long(around 7 secs)  for input dimension 700*80 with 100 BeamWidth in ARM processor which is around 5 times in comparison to x86 architecture(1.4 secs)  for the same hyperparameters. Any optimization we can do to reduce execution time in an ARM to bring it down at least equal to x86?

**Comment by githubharald** [comment]
Which mode do you use? I would suggest only using "Words" or "NGrams" mode, as they are much faster than the forecast modes, while still achieving good accuracy. Then, limit the beam width (see [README](https://github.com/githubharald/CTCWordBeamSearch#hyperparameters)): at some point, increasing it only accounts for a small accuracy improvement, while slowing down the algorithm quite a lot. Something around 30 should give a reasonable trade-off.

**Comment by iit2014128** [comment]
We are using Words mode .

**Comment by githubharald** [comment]
and beam width?

**Comment by iit2014128** [comment]
100

**Comment by githubharald** [comment]
try to go down to 30.

**Comment by githubharald** [comment]
and about which hardware are we talking? there is a wide range of ARM processors. can you give more details?

**Comment by iit2014128** [comment]
for beam width 30 it is working in 2secs but CER is going down by 5%.

**Comment by githubharald** [comment]
1. did you compile with parallel mode? https://github.com/githubharald/CTCWordBeamSearch#1-compile
if you have a batch with multiple elements, this might also improve the runtime.
2. please see my last question about hardware

**Comment by iit2014128** [comment]
yes trying to get more info about hardware, will share in a minute

**Comment by iit2014128** [comment]
ARMV7 processor rev 0(v7l)

**Comment by iit2014128** [comment]
No, We didn't compile in parallel mode.

**Comment by iit2014128** [comment]
How can we use parallel for the c++ test program?

**Comment by githubharald** [comment]
parallel mode is only implemented for TF. And it also makes sense in case a batch is processed. Do you use batches, or do you process single elements (e.g., just one input image a time)?

**Comment by iit2014128** [comment]
No, we are not using batches. we are directly feeding logits. (input to model is array of (x,y) pen co-ordinates. we are not using images).

**Comment by githubharald** [comment]
Start with the simple things:
1. search for a good beam-width, which is both fast and accurate. 
2. make sure to use some fast way to pass the data into the C++ program. CSV files are used in the test program, but not a good idea if runtime matters
3. Ideally, plug the C++ code into the main program, as I did for TF. Then, the data is directly passed from TF to the C++ program, instead of writing a file with the data

If this does not help, then there is no way around profiling the program on the hardware and searching for performance bottle-necks.

**Comment by iit2014128** [comment]
we tried profiling this program and find out that push_back in vector<vector<>> ex. wordList is taking
majority of the time. is there any alternative to this ? or any way we can make it faster?

**Comment by iit2014128** [comment]
we are creating language model only once.

**Comment by githubharald** [comment]
you said that you use Words mode. But the code newBeam->m_wordHist.push_back(newBeam->m_wordDev) is not called in words mode. Please clarify.

**Comment by iit2014128** [comment]
Sorry previously we were using words but now switched to Ngram mode.

**Comment by githubharald** [comment]
you could try to move m_wordDev instead of copying it - add a std::move and comment out the second line:

```
newBeam->m_wordHist.push_back(std::move(newBeam->m_wordDev));
//newBeam->m_wordDev.clear();
```

**Comment by githubharald** [comment]
closing because of inactivity.
